ui-admin/src/api/enrolleeSearch.ts

```diff
@@ -0,0 +1,191 @@
+    this.values = values
+    this.stableId = stableId
+  }
+
+  isDefault() {
+    return this.values.length === 0
+  }
+}
+export type StableIdStringArrayFacetValueFields = { values: StableIdStringValue[] }
+export class StableIdStringArrayFacetValue implements IFacetValue {
+  values: StableIdStringValue[]
+
+  facet: StableIdStringArrayFacet
+
+  constructor(facet: StableIdStringArrayFacet, facetVal: StableIdStringArrayFacetValueFields = { values: [] }) {
+    this.values = facetVal.values
+    this.facet = facet
+  }
+
+  isDefault() {
+    return this.values.length === 0 || !this.values.some(val => !val.isDefault())
+  }
+}
+
+export type FacetValue =  StringFacetValue | IntRangeFacetValue | StableIdStringArrayFacetValue
+
+export type FacetType = | 'INT_RANGE' | 'STRING' | 'STABLEID_STRING'
+
+export type BaseFacet = {
+  keyName: string,
+  category: string,
+  label: string,
+  type: FacetType,
+}
+
+export type IntRangeFacet = BaseFacet & {
+  type: 'INT_RANGE'
+  min: number | null
+  max: number | null
+}
+
+export type FacetOption = {
+  label: string, value: string
+}
+
+export type StringFacet = BaseFacet & {
+  type: 'STRING',
+  options: FacetOption[]
+}
+
+export type StableIdStringArrayFacet = BaseFacet & {
+  type: 'STABLEID_STRING',
+  options: FacetOption[]
+  stableIdOptions: FacetOption[]
+}
+
+export type Facet = StringFacet | StableIdStringArrayFacet | IntRangeFacet
+
+export const SAMPLE_FACETS: Facet[] = [{
+  category: 'profile',
+  keyName: 'age',
+  label: 'Age',
+  type: 'INT_RANGE',
+  max: 150,
+  min: 0
+}, {
+  category: 'profile',
+  keyName: 'sexAtBirth',
+  label: 'Sex at birth',
+  type: 'STRING',
+  options: [
+    { value: 'male', label: 'male' },
+    { value: 'female', label: 'female' },
+    { value: 'other', label: 'other' },
+    { value: 'unknown', label: 'unknown' }
+  ]
+}, {
+  category: 'participantTask',
+  keyName: 'status',
+  label: 'Task status',
+  type: 'STABLEID_STRING',
+  options: [
+    { value: 'COMPLETE', label: 'Complete' },
+    { value: 'IN_PROGRESS', label: 'In progress' },
+    { value: 'NEW', label: 'New' }
+  ],
+  stableIdOptions: [
+    { value: 'oh_oh_consent', label: 'Consent' },
+    { value: 'oh_oh_basicInfo', label: 'Basics' },
+    { value: 'oh_oh_cardioHx', label: 'Cardio History' }
+  ]
+}]
+
+
+/**
+ *
+ */
+export const newFacetValue = (facet: Facet, facetValue?: object): FacetValue => {
+  if (facet.type === 'INT_RANGE') {
+    return new IntRangeFacetValue(facet, facetValue as IntRangeFacetValueFields)
+  } else if (facet.type === 'STABLEID_STRING') {
+    const newValues = facetValue ? (facetValue as StableIdStringArrayFacetValueFields).values.map(stableIdVal =>
+      new StableIdStringValue(stableIdVal.stableId, stableIdVal.values)
+    ) : []
+    return new StableIdStringArrayFacetValue(facet, { values: newValues })
+  }
+  return new StringFacetValue(facet, facetValue as StringFacetValueFields)
+}
```

Comment:
You could have TypeScript verify that this function handles all possible values of `facet.type` with something like this:

```suggestion
const checkExhaustive = (facetType: never) => {
  throw new Error(`Unhandled facet type ${facetType}`)
}

export const newFacetValue = (facet: Facet, facetValue?: object): FacetValue => {
  const facetType = facet.type
  if (facetType === 'INT_RANGE') {
    return new IntRangeFacetValue(facet, facetValue as IntRangeFacetValueFields)
  } else if (facetType === 'STABLEID_STRING') {
    const newValues = facetValue ? (facetValue as StableIdStringArrayFacetValueFields).values.map(stableIdVal =>
      new StableIdStringValue(stableIdVal.stableId, stableIdVal.values)
    ) : []
    return new StableIdStringArrayFacetValue(facet, { values: newValues })
  } else if (facetType === 'STRING') {
    return new StringFacetValue(facet, facetValue as StringFacetValueFields)
  }
  checkExhaustive(facetType)
}
```

Reply:
You might also be able to use [overloads](https://www.typescriptlang.org/docs/handbook/2/functions.html#function-overloads) to check that the types of the facet and facetValue parameters match.

Reply:
that's a good point about there not being any validation on the types matching.  I think that we'll want that, but it won't be via typescript, since this function is almost always being called from url parsing, where the caller won't know and can't check the types.